sidechain/consensus/common/src/is_descendant_of_builder.rs

```diff
@@ -0,0 +1,132 @@
+/*
+	Copyright 2021 Integritee AG and Supercomputing Systems AG
+
+	Licensed under the Apache License, Version 2.0 (the "License");
+	you may not use this file except in compliance with the License.
+	You may obtain a copy of the License at
+
+		http://www.apache.org/licenses/LICENSE-2.0
+
+	Unless required by applicable law or agreed to in writing, software
+	distributed under the License is distributed on an "AS IS" BASIS,
+	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	See the License for the specific language governing permissions and
+	limitations under the License.
+
+*/
+use crate::header_db::HeaderDbTrait;
+use core::{hash::Hash as HashT, marker::PhantomData};
+use itp_types::H256;
+use its_primitives::traits::Header as HeaderT;
+
+pub struct IsDescendantOfBuilder<Hash, HeaderDb, Error>(PhantomData<(Hash, HeaderDb, Error)>);
+
+impl<'a, Hash, HeaderDb, Error> IsDescendantOfBuilder<Hash, HeaderDb, Error>
+where
+	Error: From<()>,
+	Hash: PartialEq + HashT + Default + Into<H256> + From<H256> + Clone,
+	HeaderDb: HeaderDbTrait,
+{
+	/// Build the `is_descendant_of` closure for the fork-tree structure
+	/// to utilize when adding and removing nodes from the tree.
+	pub fn build_is_descendant_of(
+		current: Option<(&'a Hash, &'a Hash)>,
+		header_db: &'a HeaderDb,
+	) -> impl Fn(&Hash, &Hash) -> Result<bool, Error> + 'a {
+		move |base, head| {
+			// If the base is equal to the proposed head then the head is forsure not a descendant of the base
+			if base == head {
+				return Ok(false)
+			}
+
+			let mut head = head;
+			if let Some((current_hash, current_parent_hash)) = current {
+				// If the current hash is equal to the base then it will not be a descendant of base
+				if current_hash == base {
+					return Ok(false)
+				}
+
+				// if the current hash is the head and the parent is the base then we know that
+				// this current hash is the descendant of the parent otherwise we can set the
+				// head to the parent and find the lowest common ancestor between head and base in the tree.
```

Comment:
```suggestion
				// If the current hash is the head and the parent is the base, then we know that
				// this current hash is the descendant of the parent. Otherwise we can set the
				// head to the parent and find the lowest common ancestor between `head`
				/// and `base` in the tree.
```